File: class4/js/1629.js
const fs = require("fs");
const input = fs.readFileSync("./input.txt").toString().trim().split("\n");

const divsqu = (a, b, c) => {
  if (b === 0) return 1;
  if (b === 1) return a % c;

  const half = divsqu(a, Math.floor(b / 2), c);
  let res = (half * half) % c;
  if (b % 2 === 1) res = ((res * a) % c) % c;
  return res;
};

const solution = () => {
  let [a, b, c] = input[0].split(" ").map(Number);

  console.log(divsqu(a, b, c));
};

solution();

File: class4/cpp/1629.cpp
#include <vector>
#include <string>
#include <iostream>

using namespace std;

long long divsqu(long long a, long long b, long long c){
    if (b == 0) return 1; 
    if (b == 1) return a % c;

    long long half = divsqu(a,b/2,c);
    long long res  = half*half%c;

    if(b%2 == 1){
        res = res*a;
        res = res%c;
    }
    return res;
    
}

int main (){
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    long long a, b, c;
    cin >> a >> b >> c;
    
    cout << divsqu(a,b,c) << '\n';


    return 0;
}